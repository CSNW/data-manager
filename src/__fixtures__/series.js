export function single() {
  return [
    {
      values: [
        { a: 0, b: 10 },
        { a: 1, b: 8 },
        { a: 2, b: 6 },
        { a: 3, b: 4 },
        { a: 4, b: 2 },
        { a: 5, b: 0 }
      ]
    }
  ];
}

export function multi() {
  return [single()[0], single()[0], single()[0]];
}
