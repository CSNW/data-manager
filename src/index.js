export { default as Store } from './store';
export { default as table } from './table';
export {
  filter,
  map,
  sort,
  sortBy,
  compare,
  groupBy,
  clone
} from './operations';
