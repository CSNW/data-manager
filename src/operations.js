import {
  compare,
  mapValues,
  shallowCloneObj,
  flatMap as _flatMap
} from './utils';

/**
 * Filter series values to only selected rows
 *
 * @param {function} iterator (row, index, rows) => boolean
 * @returns {function} operation
 */
export function filter(iterator) {
  return data => mapValues(data, values => values.filter(iterator));
}

/**
 * Map series values
 *
 * @param {function} iterator (row, index, rows) => row
 * @returns {function} operation
 */
export function map(iterator) {
  return data => mapValues(data, values => values.map(iterator));
}

/**
 * Map series objects
 *
 * @param {function} iterator (series, index, data) => series
 * @returns {function} operation
 */
export function mapSeries(iterator) {
  return data => data.map(iterator);
}

/**
 * Flat map series values
 *
 * @param {function} iterator (row, index, rows) => row[]
 * @returns {function} operation
 */
export function flatMap(iterator) {
  return data =>
    data.map(series => {
      series.values = _flatMap(series.values, iterator);
      return series;
    });
}

/**
 * Sort series values
 *
 * @param {function} comparator (row_a, row_b) => number
 * @returns {function} operation
 */
export function sort(comparator) {
  return data => mapValues(data, values => values.sort(comparator));
}

/**
 * Sort series values by key
 *
 * @param {string} key
 * @param {function} comparator (a, b) => number
 * @returns {function} operation
 */
export function sortBy(key, comparator) {
  return sort((row_a, row_b) => comparator(row_a[key], row_b[key]));
}

/**
 * Clone series values
 *
 * @returns {function} operation
 */
export function clone() {
  return data => mapValues(data, values => values.map(shallowCloneObj));
}

/**
 * Clone series objects (does not clone underlying rows)
 *
 * @returns {function} operation
 */
export function cloneSeries() {
  return data => data.map(series => shallowCloneObj(series));
}

/**
 * Group series values by key
 *
 * @example
 * ```js
 * import { Store, table, groupBy } from 'data-manager';
 * import cast from 'data-manager/cast.macro';
 *
 * // data.csv
 * // a,b
 * // 1,a
 * // 2,b
 * // 3,b
 * // 4,a
 * const store = new Store();
 * const data = table('data.csv', cast({ a: Number, b: String }));
 *
 * async function grouped() {
 *   const results = await store.query(
 *     data,
 *     groupBy('b')
 *   );
 *
 *   // results = [
 *   //   { group: { b: 'a' }, values: [{ a: 1, b: 'a' }, { a: 4, b: 'a' }] },
 *   //   { group: { b: 'b' }, values: [{ a: 2, b: 'b' }, { a: 3, b: 'b' }] }
 *   // ]
 *
 *   const alternative = await store.query(
 *     data,
 *     groupBy('b', (b, series) => {
 *       // `values` is added directly to returned series
 *       series.key = b;
 *
 *       return series;
 *     })
 *   );
 *
 *   // alternative = [
 *   //   { key: 'a', values: [{ a: 1, b: 'a' }, { a: 4, b: 'a' }] },
 *   //   { key: 'b', values: [{ a: 2, b: 'b' }, { a: 3, b: 'b' }] }
 *   // ]
 * }
 * ```
 * @param {string} key
 * @param {function} [toSeries]
 */
export function groupBy(key, toSeries) {
  toSeries =
    toSeries ||
    ((value, series) => {
      series.group = series.group || {};
      series.group[key] = value;

      return series;
    });

  return data =>
    _flatMap(data, series => {
      const groups = new Map();
      series.values.forEach(row => {
        const id = row[key];

        if (!groups.has(id)) groups.set(id, []);
        const group = groups.get(id);

        group.push(row);
      });

      const grouped = [];
      groups.forEach((values, id) => {
        const group_series = toSeries(id, shallowCloneObj(series));
        group_series.values = values;

        grouped.push(group_series);
      });

      return grouped;
    });
}
