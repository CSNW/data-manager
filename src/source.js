export default class Source {
  fetch(path, map) {
    // TODO
  }

  query() {
    // TODO
  }
}
