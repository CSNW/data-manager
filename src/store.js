export default class Store {
  constructor(overrides = {}) {
    const { fetch = fetchCsv } = overrides;
    this._fetch = fetch;
  }

  fetch(path, map) {
    return Promise.resolve(this._fetch(path)).then(data => {
      return map ? data.map(map) : data;
    });
  }

  query(...operations) {
    return operations.reduce((loading, operation) => {
      return loading.then(data => operation(data, this));
    }, Promise.resolve([]));
  }
}

function fetchCsv(path) {
  // TODO
  return [];
}
