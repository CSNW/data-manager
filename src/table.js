export default function table(path, map) {
  return (_, store) => {
    return store.fetch(path, map).then(values => [{ values }]);
  };
}
