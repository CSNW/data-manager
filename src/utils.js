export function mapValues(data, iterator) {
  return data.map(series => {
    const mapped = shallowCloneObj(series);
    mapped.values = iterator(series.values);

    return mapped;
  });
}

export function shallowCloneObj(obj) {
  const cloned = {};
  for (const key in obj) {
    cloned[key] = obj[key];
  }
  return cloned;
}
