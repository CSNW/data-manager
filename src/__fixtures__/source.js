import Source from '../source';
import { single } from './series';

export default function source() {
  const override = new Source();
  override.fetch = jest.fn(() => Promise.resolve(single()[0].values));

  return override;
}
