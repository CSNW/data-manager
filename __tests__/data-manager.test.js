import { Store, table, filter } from '../src/index';
import { single } from '../src/__fixtures__/series';

test('should fetch and filter csv data', async () => {
  const store = new Store({
    fetch() {
      return single()[0].values;
    }
  });

  const results = await store.query(
    table('data.csv'),
    filter(row => row.a > 2)
  );

  expect(results).toMatchSnapshot();
});
