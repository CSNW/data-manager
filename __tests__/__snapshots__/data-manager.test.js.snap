// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch and filter csv data 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": 3,
        "b": 4,
      },
      Object {
        "a": 4,
        "b": 2,
      },
      Object {
        "a": 5,
        "b": 0,
      },
    ],
  },
]
`;
